src/DotNetty.Common/ThreadLocalPool.cs

```diff
@@ -236,83 +236,109 @@ internal void Add(DefaultHandle handle)
-                            throw new InvalidOperationException("recycled already");
+                            DefaultHandle element = srcElems[i];
+                            if (element == null)
+                            {
+                                return false;
+                            }
+
+                            if (element.recycleId == 0)
+                            {
+                                element.recycleId = element.lastRecycledId;
+                            }
+                            else if (element.recycleId != element.lastRecycledId)
+                            {
+                                throw new InvalidOperationException("recycled already");
+                            }
+
+                            srcElems[i] = null;
+
+                            if (dst.DropHandle(element))
+                            {
+                                // Drop the object.
+                                continue;
+                            }
+
+                            element.Stack = dst;
+                            dstElems[newDstSize++] = element;
                         }
-                        srcElems[i] = null;
 
-                        if (dst.DropHandle(element))
+                        if (srcEnd == LinkCapacity && head.next != null)
                         {
-                            // Drop the object.
-                            continue;
+                            // Add capacity back as the Link is GCed.
+                            this.head.ReclaimSpace(LinkCapacity);
+                            this.head.link = head.next;
                         }
-                        element.Stack = dst;
-                        dstElems[newDstSize++] = element;
-                    }
 
-                    if (srcEnd == LinkCapacity && head.next != null)
-                    {
-                        // Add capacity back as the Link is GCed.
-                        this.head.ReclaimSpace(LinkCapacity);
-                        this.head.link = head.next;
-                    }
+                        head.ReadIndex = srcEnd;
+                        if (dst.size == newDstSize)
+                        {
+                            return false;
+                        }
 
-                    head.ReadIndex = srcEnd;
-                    if (dst.size == newDstSize)
+                        dst.size = newDstSize;
+                        return true;
+                    }
+                    else
                     {
+                        // The destination stack is full already.
                         return false;
                     }
-                    dst.size = newDstSize;
-                    return true;
                 }
-                else
+                catch (NullReferenceException)
                 {
-                    // The destination stack is full already.
                     return false;
                 }
             }
```

Comment:
Would skip this part for now, let's see if null checks helped.

Reply:
ok, done